File: app/src/main/java/net/wojteksz128/worktimemeasureapp/util/PeriodicOperationRunner.java
package net.wojteksz128.worktimemeasureapp.util;

import android.util.Log;

import java.util.concurrent.atomic.AtomicBoolean;

public class PeriodicOperationRunner<T> {

    private int sleepDuration = 1000;
    private Consumer<T> consumer;
    private final AtomicBoolean running = new AtomicBoolean(false);
    private Thread thread;


    public PeriodicOperationRunner() {}

    public void start() {
        if (thread == null || !isRunning()) {
            this.thread = new PeriodicRun();
            thread.start();
        }
    }

    public void stop() {
        running.set(false);
    }

    @SuppressWarnings("BooleanMethodIsAlwaysInverted")
    public boolean isRunning() {
        return running.get();
    }

    public void setConsumer(Consumer<T> consumer) {
        this.consumer = consumer;
    }

    private class PeriodicRun extends Thread {

        private final String TAG = PeriodicOperationRunner.PeriodicRun.class.getSimpleName();

        @Override
        public void run() {
            PeriodicOperationRunner.this.running.set(true);
            while (running.get()) {
                try {
                    Thread.sleep(PeriodicOperationRunner.this.sleepDuration);
                    if (PeriodicOperationRunner.this.consumer != null && running.get()) {
                        PeriodicOperationRunner.this.consumer.invoke();
                    } else {
                        Log.w(TAG, "run: consumer is null or running is ending");
                    }
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }
    }
}

File: app/src/main/kotlin/net/wojteksz128/worktimemeasureapp/util/PeriodicOperationRunner.kt
package net.wojteksz128.worktimemeasureapp.util

import android.util.Log

import java.util.concurrent.atomic.AtomicBoolean


class PeriodicOperationRunner<T> {

    private val sleepDuration = 1000
    private var functionWithParameter: FunctionWithParameter<T>? = null
    private val running = AtomicBoolean(false)
    private var thread: Thread? = null

    val isRunning: Boolean
        get() = running.get()

    fun start() {
        if (thread == null || !isRunning) {
            this.thread = PeriodicRun()
            thread!!.start()
        }
    }

    fun stop() {
        running.set(false)
    }

    fun setConsumer(functionWithParameter: FunctionWithParameter<T>) {
        this.functionWithParameter = functionWithParameter
    }

    private inner class PeriodicRun : Thread() {

        @Suppress("PrivatePropertyName")
        private val TAG = PeriodicRun::class.java.simpleName

        override fun run() {
            running.set(true)
            while (running.get()) {
                try {
                    sleep(sleepDuration.toLong())
                    if (functionWithParameter != null && running.get()) {
                        functionWithParameter!!.invoke()
                    } else {
                        Log.w(TAG, "run: functionWithParameter is null or running is ending")
                    }
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }

            }
        }
    }
}
